# NumPy - Manipulando Dados Numéricos

In [ ]:
# NumPy na Prática
import numpy as np             # Importa a biblioteca NumPy para seu projeto

vetor = np.array([0, 1, 2, 3]) # instancia um vetor (array) em NumPy
vetor                          # Imprime seu conteúdo

| Tipo de Dado | Descrição                                            |
|:-------------|:-----------------------------------------------------|
| np.int64     | Números inteiros de 64bits                           |
| np.float32   | Números de ponto flutuante de precisão dupla (Reais) |
| np.complex   | Números Complexos                                    |
| np.bool      | Tipo lógico (Booleano)                               |
| np.object    | Tipo object do Python                                |
| np.string_   | Cadeia de Caracteres Fixa (String)                   |
| np.unicode_  | Cadeia de Caracteres Unicode Fixa (String)           |

### Construção Manual de Vetores

In [ ]:
# CONSTRUÇÃO MANUAL DE VETORES
# Unidimencional (1D)

vetor_numpy = np.array([0, 1, 2, 3, 4])

print(vetor_numpy)        # Imprime o vetor
print(vetor_numpy.ndim)   # Retorna o número de dimensões do vetor
print(vetor_numpy.shape)  # Retorna a forma do vetor ou atribui uma tupla para as dimnesões do vetor
print(len(vetor_numpy))   # Retorna o número de elementos do vetor

# Multi-dimencional (n-D)

# Matriz Bidimensional 2 (linhas) X 3 (colunas)
matriz_numpy = np.array([[0, 1, 2], [3, 4, 5]]) 

print(matriz_numpy)
print(matriz_numpy.ndim)
print(matriz_numpy.shape)
print(len(matriz_numpy))         # Retorna o tamanho da primeira dimensão do vetor (linhas)

### Funções para Criação de Vetores

In [ ]:
# Vetores Igualmente Espaçado
print("Vetores Igualmente Espaçado")
vetor = np.arange(10) # 0 .. n - 1
print(vetor)

vetor_impar = np.arange(1, 9, 2) # Início, Fim e Incremento
print(vetor_impar)

# Vetores por números de pontos
print("\nVetores por Números de Pontos")
numero_pontos = np.linspace(0, 1, 6) # Início, Fim, Número de Pontos
print(numero_pontos)

# Matrizes Populares
print("\nMATRIZES POPULARES")
print("\nMatriz de Uns")
matriz_de_uns = np.ones((3, 3)) # Tupla (3, 3)
print(matriz_de_uns)

print("\nMatriz de Zeros")
matriz_de_zeros = np.zeros((2, 2))
print(matriz_de_zeros)

print("\nMatriz Zerada com Diagonal de Uns")
matriz_zeros_diagonal_uns = np.eye(4)
print(matriz_zeros_diagonal_uns)

print("\nMatriz Zerada com Diagonal de Permutações")
matriz_diagonal_permutacao = np.diag(np.array([1, 2, 3, 4]))
print(matriz_diagonal_permutacao)

print("\nVETORES ALEATÓRIOS")
print("\nVetor Aleatório Uniforme")
vetor_uniforme_aleatorio = np.random.rand(4)
print(vetor_uniforme_aleatorio)

print("\nVetor Aleatório Gaussiano")
vetor_aleatorio_gaussiano = np.random.randn(4)
print(vetor_aleatorio_gaussiano)

np.random.seed(1234) # Definindo a semente aleatória

### Operações Matemáticas com Vetores

In [ ]:
import numpy as np

vetor_a = np.array([1, 2, 3])
vetor_b = np.array([4, 5, 6])


# Adição
vetor_a + vetor_b              # Usando operador aritmético
np.add(vetor_a, vetor_b)       # Usando função definida do NumPy

# Subtração
vetor_a - vetor_b              # Usando operador aritmético
np.subtract(vetor_a, vetor_b)  # Usando função definida do NumPy

# Multiplicação
vetor_a * vetor_b              # Usando operador aritmético
np.multiply(vetor_a, vetor_b)  # Usando função definida do NumPy

# Divisão
vetor_a / vetor_b              # Usando operador aritmético
np.divide(vetor_a, vetor_b)    # Usando função definida do NumPy

# Raiz Quadrada
np.sqrt(vetor_a)

# Função Exponencia
np.exp(vetor_a)                

# Seno
np.sin(vetor_a)

# Coseno
np.cos(vetor_a)

# Logaritmo Natural
np.log(vetor_a)

### Comparação

In [ ]:
import numpy as np

vetor_a = np.array([1, 2, 3])
vetor_b = np.array([4, 5, 6])

# Compara cada um dos elementos dos vetores
# Retorna um vetor booleano com os resultados da comparação
vetor_a == vetor_b
vetor_a != vetor_b

vetor_a > vetor_b
vetor_a >= vetor_b

vetor_a < vetor_b
vetor_a <= vetor_b

### Funções Agregadas

In [ ]:
import numpy as np

vetor = np.array([1, 2, 3, 4, 5])

vetor.sum()         # Soma todos os elementos do vetor
vetor.cumsum()      # Soma acumulada de todos os elementos do vetor

vetor.min()         # Retorna o menor elemento do vetor numérico
vetor.max()         # Retorna o maior elemento do vetor numérico

vetor.mean()        # Média
np.median(vetor)    # Mediana
np.corrcoef(vetor)  # Coeficiente de Correlação
np.std(vetor)       # Desvio Padrão

### Copiando Vetores

In [ ]:
import numpy as np

vetor = np.array([1, 2, 3, 4, 5])

copia_vetor = np.copy(vetor)
print(copia_vetor)

copia = vetor.copy()  
print(copia)

visao = vetor.view()  # Cria uma visão (view) do vetor [representação virutal em memória]
print(visao)

### Ordenando Vetores

In [ ]:
import numpy as np

vetor = np.array([9, 8, 7, 6, 5, 4, 3, 2, 1])
vetor.sort()
vetor

### Subconjunto, Fatiamento e Indexação

In [ ]:
import numpy as np

vetor = np.array([1, 2, 3, 4, 5])
matriz = np.array([(1, 2, 3), (4, 5, 6), (7, 8, 9)])


print(vetor[2])        # vetor[i] = Pega o elemento do vetor na posição (i)
print(matriz[1, 2])    # vetor[linha, coluna] = Pega um elemento em uma matriz


print(vetor[0:3])      # vetor[inicio:quantidade] = Retorna 3 elementos contados à partir da posição de início ([0] 
print(matriz[1:2])     # Seleciona todos elementos da primeira linha até a terceira coluna
print(matriz[0:3, 1])  # matriz[inicio:quantidade, coluna] Seleciona 3 elementos da segunda coluna [1], à partir da primeira linha [0]

print(matriz[:1])      # Seleciona todos os itens da primeira linha
print(matriz[:2])      # Seleciona todos os itens da primeira linha e segunda linha

print(vetor[ : :-1])   # Seleciona todos os elementos de trás pra frente (invertido)

# Indexação Lógica
print(vetor[vetor <= 3]) # Seleciona todos os elementos dos índices menores iguais a 3
print(vetor[1 + 3])      # Calcula dinamicamente a posição do vetor


print(matriz[[1, 0, 1, 0]]) # Retorna uma matriz com as linhas 1 e 0 repetidas

In [ ]:
import numpy as np
from io import StringIO

data = u"1, 2, 3\n4, 5, 6"
vetor = np.genfromtxt(StringIO(data), delimiter=",") # Converte dados para vetor
np.savetxt("nome_arquivo.txt", vetor, delimiter=" ") # Salva o vetor como arquivo

```python
# Exemplo - Carregamento de Arquivos
np.genfromtxt("nome_arquivo.csv", delimiter=',')
np.loadtxt("nome_arquivo.txt")
```

# Exercícios

1. Importe o pacote numpy com o nome np.

In [2]:
import numpy as np

2. Crie o seguinte vetor unidimensional (1D):

| 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 |
|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|

In [3]:
vet = np.array(range(10))
vet

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

3. Qual é o comprimento `len` do vetor?

In [4]:
len(vet)

10

4. Quantas dimensões `ndim` o vetor possui?

In [5]:
vet.ndim

1

5. O que a função `shape` retorna?

In [6]:
vet.shape

(10,)

**Resposta**: A função shape retorna o formato do vetor, o que indica seu tamanho. No caso acima, existem 10 registros, como se trata de um vetor de apenas uma dimensão, é retornado apenas o número de elementos.

6. Crie a variável *pares* e extraia todos os números pares do vetor para ela:

In [7]:
pares = vet[0::2]
pares

array([0, 2, 4, 6, 8])

7. Substitua todos os elementos ímpares do vetor por -1, em seguida, ordene os resultados:

In [8]:
vet[vet % 2 != 0] = -1
vet.sort()

vet

array([-1, -1, -1, -1, -1,  0,  2,  4,  6,  8])

8. Converta o vetor unidimensional (1D) abaixo para uma matriz (2D) com 4 linhas. **Dica:** `reshape`.

In [15]:
vetor1d = np.arange(12)
vetor2d = vetor1d.reshape(4,3)

vetor2d

array([[ 0,  1,  2],
       [ 3,  4,  5],
       [ 6,  7,  8],
       [ 9, 10, 11]])

9. Obtenha os itens em comum, ou seja, a intersecção dos vetores `a` e `b` abaixo:
**Saída**: [2, 4, 6]

In [16]:
a = np.array([1, 2, 3, 4, 5, 6])
b = np.array([2, 4, 6, 8, 10])

commonItems = np.intersect1d(a, b)
commonItems

array([2, 4, 6])

10. Retorne os valores únicos do vetor `a` que não estão no vetor `b` (diferença).
**Saída**: [1, 3]

In [21]:
a = np.array([1, 2, 3, 4])
b = np.array([2, 4, 6, 8, 10])

diff = np.setdiff1d(a,b)
diff

array([1, 3])

11. Retorne um vetor com as posições (índices) onde os elementos do vetor `a` sãos iguais aos elementos do vetor `b`. **Dica**: `where`.
**Saída**: [2, 3, 5]

In [42]:
a = np.array([1, 2, 3, 4, 5, 6])
b = np.array([2, 4, 3, 4, 10, 6])

indices = np.where(a == b)
indices

(array([2, 3, 5], dtype=int64),)

12. Obtenha todos os itens entre 5 e 15 do vetor `a = [0, 1, 2, 3, . . ., 20]`.

In [61]:
a = np.arange(21)
interval = np.logical_and(a >= 5, a <= 15)
a[interval]

array([ 5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15])

13. Crie e imprima um vetor com 5 números aleatórios, sem seguida, obtenha os valores máximo e mínimo dele e mostre na tela.

In [70]:
rand = np.random.randint(1, 100, 5)
print(f'Vetor de 5 números aleatórios: {rand}')
print(f'\n- Maior valor do vetor: {rand.max()}')
print(f'- Menor valor do vetor: {rand.min()}')

Vetor de 5 números aleatórios: [65  4 42  1  8]

- Maior valor do vetor: 65
- Menor valor do vetor: 1


14. Crie a seguinte matriz:

|   |   |   |
|:-:|:-:|:-:|
| 7 | 8 | 9 |
| 4 | 5 | 6 |
| 1 | 2 | 3 |

In [80]:
matriz = np.array([[7,8,9],[4,5,6], [1,2,3]])
matriz

array([[7, 8, 9],
       [4, 5, 6],
       [1, 2, 3]])

15. Troque de lugar a primeira coluna com a útltima coluna da matriz.

In [81]:
matrizCopy = matriz.copy()
matriz[:,0] = matrizCopy[:,-1]
matriz[:,-1] = matrizCopy[:,0]

matriz

array([[9, 8, 7],
       [6, 5, 4],
       [3, 2, 1]])

16. Troque de lugar a primeira linha com a útltima linha da matriz.

In [82]:
matrizCopy = matriz.copy()
matriz[0,:] = matrizCopy[-1,:]
matriz[-1,:] = matrizCopy[0,:]

matriz

array([[3, 2, 1],
       [6, 5, 4],
       [9, 8, 7]])

17. Crie uma matriz  7 × 5 preenchida com números decimais aleatórios entre 1 e 5.

In [115]:
values = []
for i in range(7):
    values.append(np.random.uniform(0.0, 5.0, (5,)))

randomMatriz = np.array(values)
randomMatriz

array([[4.84906031, 0.03873704, 1.37117428, 0.81146773, 3.97487712],
       [1.50134872, 2.27049715, 2.0631212 , 3.93706101, 0.72377331],
       [2.24848756, 3.63993323, 0.12614414, 4.21112462, 0.88280815],
       [2.12192913, 3.81892417, 4.08008441, 2.55197622, 4.64441423],
       [3.98987993, 0.21197132, 1.13417161, 1.01703378, 2.24712654],
       [1.94725332, 1.16234323, 1.98112167, 4.76405021, 3.50632694],
       [2.2667991 , 3.92247722, 2.3534885 , 4.38884688, 2.33837067]])

18. Qual é o comprimento `len` da matriz?

In [116]:
len(randomMatriz)

7

19. Quantas dimensões `ndim` a matriz possui?

In [117]:
np.ndim(randomMatriz)

2

20. O que a função `shape` retorna?

In [118]:
np.shape(randomMatriz)

(7, 5)

**Resposta**: Retorna o formato do vetor. Neste caso, o vetor possui 7 linhas, cada uma possuindo 5 colunas

# EXERCÍCIO: Análise de Dados dos Jogadores da FIFA ⚽

Neste exercício, iremos usar a biblioteca NumPy para fazer uma análise de dados de jogares de futebol da FIFA disponíveis em [`data/fifa.csv`](data/fifa.csv). O conjunto de dados (dataset) foi adaptado da versão original disponibilizada nas plataformas: [Kaggle](https://www.kaggle.com/karangadiya/fifa19/data) e [GitHub](https://github.com/amanthedorkknight/fifa18-all-player-statistics).

## Dicionário de Dados

| Coluna        | Descrição                                        |
|:--------------|:-------------------------------------------------|
| id            | Identificador único do jogador                   |
| nome          | Nome do jogador                                  |
| foto          | URL da fotografia do jogador.                    |
| nacionalidade | País do jogador.                                 |
| avaliacao     | Avaliação geral do jogador.                      |
| potencial     | Indicador da classificação potencial do jogador. |
| time          | Nome do time do jogador.                         |
| pe_preferido  | Direito (Destro), Esquerdo (Canhoto).            |
| reputacao     | Reputação Internacional do jogador.              |
| altura_cm     | Altura do jogador em centímetros (cm).           |
| peso_kg       | Peso do jogador em kilos (kg).                   |
| posicao       | Posição do jogador no time.                      |
| numero_camisa | Número da camisa do jogador.                     |

1. Importe o pacote numpy com o nome np.

In [1]:
import numpy as np

2. Carregue o conjunto de dados da Fifa disponível na pasta (`data/fifa.csv`), ignorando a coluna `foto`, para uma matriz em uma variavel chamada `dados`. **Dicas:**
    - Pesquise pela função `genfromtxt`; 
    - Não esqueça de especificar a codificação (encoding) como UTF-8;
    - Ignore o cabeçalho dos dados usando o parâmetro `skip_header=True`

In [178]:
# remoção das colunas com links para imagens
dados = np.genfromtxt('data/fifa18.csv', encoding='utf8', dtype=object, delimiter=',', skip_header=True, usecols=(0,1,2,4,5,6,7,8,9,10,11,12,13))

#remocao dos jogadores sem algum tipo de registro dos dados físicos
empty_lines = np.where(dados[:, 9:12] == b'')
dados = np.delete(dados, empty_lines[0], axis=0)

dados

array([[b'158023', b'L. Messi', b'31', ..., b'72.121', b'RF', b'10'],
       [b'20801', b'Cristiano Ronaldo', b'33', ..., b'83.007', b'ST',
        b'7'],
       [b'190871', b'Neymar Jr', b'26', ..., b'68.039', b'LW', b'10'],
       ...,
       [b'241638', b'B. Worman', b'16', ..., b'67.132', b'ST', b'33'],
       [b'246268', b'D. Walker-Rice', b'17', ..., b'69.853', b'RW',
        b'34'],
       [b'246269', b'G. Nugent', b'16', ..., b'79.832', b'CM', b'33']],
      dtype=object)

3. Quantas linhas foram lidas do arquivo?

In [179]:
len(dados)

18147

4. Quantas dimensões sua variável `dados` possui?

In [144]:
np.ndim(dados)

2

5. O que a função shape retorna?

In [145]:
np.shape(dados)

(18147, 13)

**Resposta**: 18147 linhas com 13 colunas cada

6. Mostre o dado que está na 3ª linha da 4ª coluna?

In [104]:
dados[2,3]

b'Brasil'

7. Exiba os dados da primeira linha do dataset.

In [105]:
dados[0]

array([b'158023', b'L. Messi', b'31', b'Argentina', b'94', b'94',
       b'FC Barcelona', b'Esquerdo', b'5', b'170', b'72.121', b'RF',
       b'10'], dtype=object)

8. Mostre as 5 primeiras linhas do conjunto de dados.

In [146]:
dados[0:5, :]

array([[b'158023', b'L. Messi', b'31', b'Argentina', b'94', b'94',
        b'FC Barcelona', b'Esquerdo', b'5', b'170', b'72.121', b'RF',
        b'10'],
       [b'20801', b'Cristiano Ronaldo', b'33', b'Portugal', b'94', b'94',
        b'Juventus', b'Direito', b'5', b'188', b'83.007', b'ST', b'7'],
       [b'190871', b'Neymar Jr', b'26', b'Brasil', b'92', b'93',
        b'Paris Saint-Germain', b'Direito', b'5', b'175', b'68.039',
        b'LW', b'10'],
       [b'193080', b'De Gea', b'27', b'Espanha', b'91', b'93',
        b'Manchester United', b'Direito', b'4', b'193', b'76.204', b'GK',
        b'1'],
       [b'192985', b'K. De Bruyne', b'27', b'B\xe9lgica', b'91', b'92',
        b'Manchester City', b'Direito', b'4', b'155', b'69.853', b'RCM',
        b'7']], dtype=object)

9. Selecione todas as linhas da coluna `nome` (segunda coluna). **Dica:** use o operador `:`

In [147]:
dados[:,1]

array([b'L. Messi', b'Cristiano Ronaldo', b'Neymar Jr', ..., b'B. Worman',
       b'D. Walker-Rice', b'G. Nugent'], dtype=object)

10. Inicialize um vetor de inteiros NumPy chamado: **idades**; com todas as linhas da coluna `idade` (3ª coluna).

In [148]:
idades = dados[:,2].astype('int8')
idades

array([31, 33, 26, ..., 16, 17, 16], dtype=int8)

11. Qual é a média de idades dos jogadores?

In [149]:
mediaIdade = np.mean(idades)
mediaIdade

25.12112194853144

12. Qual é a maior idade do conjunto de dados?

In [110]:
maiorIdade = np.max(idades)
maiorIdade

45

13. Qual é a menor idade do conjunto de dados?

In [150]:
menorIdade = np.min(idades)
menorIdade

16

14. Selecione todos os jogadores que possuem a maior idade deste conjunto de dados. Quantas linhas retornaram? 

In [151]:
velhos = dados[dados[:,2].astype('int8') == maiorIdade]
velhos

array([[b'140029', b'O. P\xe9rez', b'45', b'', b'71', b'71', b'Pachuca',
        b'Direito', b'2', b'173', b'74.843', b'GK', b'21']], dtype=object)

15. Selecione todos os jogadores que possuem a menor idade deste conjunto de dados. Quantas linhas retornaram? 

In [152]:
novos = dados[dados[:,2].astype('int8') == menorIdade]
novos

array([[b'241266', b'W. Geubbels', b'16', b'', b'64', b'86',
        b'AS Monaco', b'Direito', b'1', b'185', b'72.121', b'ST', b'13'],
       [b'244403', b'A. Taoui', b'16', b'', b'64', b'75',
        b'Toulouse Football Club', b'Esquerdo', b'1', b'152', b'68.946',
        b'ST', b'35'],
       [b'245616', b'Pelayo Morilla', b'16', b'', b'63', b'81',
        b'Real Sporting de Gij\xf3n', b'Esquerdo', b'1', b'183',
        b'69.853', b'RM', b'31'],
       [b'246465', b'Guerrero', b'16', b'', b'63', b'76',
        b'CF Rayo Majadahonda', b'Esquerdo', b'1', b'155', b'74.843',
        b'CM', b'26'],
       [b'246594', b'H. Massengo', b'16', b'', b'62', b'75',
        b'AS Monaco', b'Direito', b'1', b'152', b'66.224', b'CDM', b'42'],
       [b'246419', b'Y. Verschaeren', b'16', b'', b'62', b'83',
        b'RSC Anderlecht', b'Direito', b'1', b'173', b'63.049', b'CAM',
        b'51'],
       [b'245015', b'Y. Roemer', b'16', b'', b'59', b'75', b'VVV-Venlo',
        b'Direito', b'1', b'155', b'

16. Selecione todos os jogadores brasileiros para uma variável chamada dados_brasil. Use `b'Brasil'` para fazer a comparação.

In [153]:
brasileiros = dados[dados[:,3] == b'Brasil']
brasileiros

array([[b'190871', b'Neymar Jr', b'26', b'Brasil', b'92', b'93',
        b'Paris Saint-Germain', b'Direito', b'5', b'175', b'68.039',
        b'LW', b'10'],
       [b'200145', b'Casemiro', b'26', b'Brasil', b'88', b'90',
        b'Real Madrid', b'Direito', b'3', b'185', b'83.915', b'CDM',
        b'14'],
       [b'189242', b'Coutinho', b'26', b'Brasil', b'88', b'89',
        b'FC Barcelona', b'Direito', b'3', b'173', b'68.039', b'LW',
        b'7'],
       [b'176676', b'Marcelo', b'30', b'Brasil', b'88', b'88',
        b'Real Madrid', b'Esquerdo', b'4', b'175', b'79.832', b'LB',
        b'12'],
       [b'164240', b'Thiago Silva', b'33', b'Brasil', b'88', b'88',
        b'Paris Saint-Germain', b'Direito', b'4', b'183', b'82.100',
        b'RCB', b'2'],
       [b'210257', b'Ederson', b'24', b'Brasil', b'86', b'90',
        b'Manchester City', b'Esquerdo', b'2', b'188', b'86.183', b'GK',
        b'31'],
       [b'201942', b'Roberto Firmino', b'26', b'Brasil', b'86', b'87',
        b'Liver

17. Quantos são os jogadores brasileiros?

In [154]:
len(brasileiros)

69

18. Para as colunas `idade`, `altura` e `peso`, calcule:
    - Média
    - Mediana
    - Máximo
    - Mínimo

In [165]:
idades = dados[:,2].astype('int8')
alturas = dados[:,9].astype('float16')
pesos = dados[:,10].astype('float16')

print(f'- Média das idades: {idades.mean()}')
print(f'- Mediana das idades: {np.median(idades)}')
print(f'- Maior idade: {idades.max()}')
print(f'- Menor idade: {idades.min()}')

print(f'\n- Média das alturas: {alturas.mean()} cm')
print(f'- Mediana das alturas: {np.median(alturas)} cm')
print(f'- Maior altura: {alturas.max()} cm')
print(f'- Menor altura: {alturas.min()} cm')

print(f'\n- Média dos pesos: {pesos.mean()} kg')
print(f'- Mediana dos pesos: {np.median(pesos)} kg')
print(f'- Maior peso: {pesos.max()} kg')
print(f'- Menor peso: {pesos.min()} kg')

- Média das idades: 25.12112194853144
- Mediana das idades: 25.0
- Maior idade: 45
- Menor idade: 16

- Média das alturas: 174.75 cm
- Mediana das alturas: 175.0 cm
- Maior altura: 206.0 cm
- Menor altura: 152.0 cm

- Média dos pesos: 75.3125 kg
- Mediana dos pesos: 74.8125 kg
- Maior peso: 110.25 kg
- Menor peso: 49.90625 kg


19. Calcule o desvio padrão da `reputação` dos jogadores brasileiros

In [169]:
brazilianRep = dados[dados[:,3] == b'Brasil', 8].astype('int8')
desvPad = np.std(brazilianRep)

desvPad

1.0093034983466755

20. Qual jogador brasileiro tem menor pontencial?

In [176]:
brazilianPots = dados[dados[:,3] == b'Brasil', 5].astype('int8')
menorBraPot =  dados[np.logical_and(dados[:,3] == b'Brasil', dados[:,5].astype('int8') == brazilianPots.min())]

menorBraPot

array([[b'230498', b'Luimo Boas Santos', b'30', b'Brasil', b'79', b'79',
        b'Santos', b'Direito', b'1', b'175', b'73.028', b'ST', b'11'],
       [b'230312', b'Gabri Prest\xe3o', b'26', b'Brasil', b'79', b'79',
        b'Internacional', b'Direito', b'1', b'185', b'79.832', b'GK',
        b'12'],
       [b'230251', b'Melvin Parrela', b'30', b'Brasil', b'79', b'79',
        b'Am\xe9rica FC (Minas Gerais)', b'Direito', b'1', b'188',
        b'87.997', b'LCM', b'14'],
       [b'230212', b'Ant\xf4nio Chiamuloira', b'30', b'Brasil', b'79',
        b'79', b'Internacional', b'Esquerdo', b'1', b'183', b'78.925',
        b'CB', b'15'],
       [b'230201', b'Maikel Catarino', b'34', b'Brasil', b'79', b'79',
        b'Atl\xe9tico Mineiro', b'Direito', b'1', b'152', b'74.843',
        b'RB', b'2']], dtype=object)